src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,429 @@
+                throw new InvalidOperationException($"The Toffoli simulator can only perform controlled rotations of multiples of 2*pi.");
+            }
+        }
+
+        public override void Exp(IQArray<Pauli> bases, double theta, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (CheckRotation(pauli, theta).isX) {
+                    X(qubit);
+                }
+            }
+        }
+
+        public override void ControlledExp(IQArray<Qubit> controls, IQArray<Pauli> bases, double theta, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (!CheckRotation(pauli, theta).safe) {
+                   throw new InvalidOperationException($"The Toffoli simulator can only perform controlled rotations of multiples of 2*pi.");
+                }
+            }
+        }
+
+        public override void ExpFrac(IQArray<Pauli> bases, long numerator, long power, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (CheckRotation(pauli, numerator, power).isX) {
+                    X(qubit);
+                }
+            }
+        }
+
+        public override void ControlledExpFrac(IQArray<Qubit> controls, IQArray<Pauli> bases, long numerator, long power, IQArray<Qubit> qubits) {
+            if (qubits.Length != bases.Length) {
+                throw new InvalidOperationException($"Both input arrays for Exp (bases, qubits), must be of same size.");
+            }
+
+            foreach (var (qubit, pauli) in qubits.Zip(bases, (q, p) => (q, p))) {
+                if (!CheckRotation(pauli, numerator, power).safe) {
+                   throw new InvalidOperationException($"The Toffoli simulator can only perform controlled rotations of multiples of 2*pi.");
+                }
+            }
+        }
+
+        // If the simulation value of qubit is true, when it's being reset, we
+        // restore it to false.
+        public override void Reset(Qubit qubit) {
+            SetValue(qubit, false);
+        }
+
+        // Measuring the qubit corresponds to translating the current simulation
+        // value into a Result value.
+        public override Result M(Qubit qubit) =>
+            GetValue(qubit).ToResult();
```

Comment:
For consistency reasons with other simple methods in this file, I'd suggest to use the standard function syntax rather than a lambda.

Reply:
I think it's consistent for single-line non-void functions to use expression syntax, see e.g., https://github.com/microsoft/qsharp-runtime/blob/master/src/Simulation/EntryPointDriver/Parsers.cs

Reply:
I am not sure about the recommended style in those cases but I find it weird (and harder to read) when the syntax keeps switching between function and expression styles like it does here for Reset and M, both of which are single line (Parsers.cs doesn't contain any void single line functions).

---
In reply to: [444431113](https://github.com/microsoft/qsharp-runtime/pull/276#discussion_r444431113) [](ancestors = 444431113)

Reply:
Please do not use lambdas unless absolutely necessary (which is not the case here).  Please use simple concepts instead of complex ones in such simple cases.

---
In reply to: [444449794](https://github.com/microsoft/qsharp-runtime/pull/276#discussion_r444449794) [](ancestors = 444449794,444431113)